src/implementation/MassaWalletProviders.ts

```diff
@@ -0,0 +1,106 @@
+import { uid } from 'uid';
+import { ICustomEventMessageResponse } from '../interfaces/ICustomEventMessageResponse';
+import { ICustomEventMessageRequest } from '../interfaces/ICustomEventMessageRequest';
+import { AvailableCommands } from './Commands';
+import { IRegisterEvent } from '../interfaces';
+
+const MASSA_WINDOW_OBJECT_PRAEFIX = 'massaWalletProvider';
+
+type CallbackFunctionVariadicAnyReturn = (
+  error: Error | null,
+  result: any,
+) => any;
+
+// =========================================================
+
+export class MassaWalletProviders {
+  private static instance: MassaWalletProviders;
+  private registeredProviders: { [key: string]: string } = {};
+  private pendingRequests: Map<string, CallbackFunctionVariadicAnyReturn>;
+
+  public static init(): MassaWalletProviders {
+    if (!MassaWalletProviders.instance) {
+      MassaWalletProviders.instance = new MassaWalletProviders();
+    }
+
+    return MassaWalletProviders.instance;
+  }
+
+  private constructor() {
+    this.getWalletProviders = this.getWalletProviders.bind(this);
+    this.handleResponseFromContentScript =
+      this.handleResponseFromContentScript.bind(this);
+    this.sendMessageToContentScript =
+      this.sendMessageToContentScript.bind(this);
+
+    this.pendingRequests = new Map<string, CallbackFunctionVariadicAnyReturn>();
+
+    // global event target to use for all wallet provider
+    window.massaWalletProvider = new EventTarget();
+
+    // hook up register handler
+    (window.massaWalletProvider as EventTarget).addEventListener(
```

Comment:
You've go two way to do it:
 1. directly using window => send will need to use [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to send event.

2. add an invisible HTML element and set a listner to it like the following
```js
      const inv = document.createElement("p");
      inv.id = "massaWalletProvider-testGreg";
      inv.style = "display:none";
      document.body.appendChild(inv);
      document.getElementById("massaWalletProvider-testGreg").addEventListener(
        'message',
        function(event) {
          console.log(event);
        }
      );
```

From my personal point of view, both scenarios are equivalent, even if the warning at the end of this section can be scary in case of an audit: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard

Reply:
I think the `document` is not shared between app script and content script, in that case the only option is postMessage

Reply:
@gregLibert Both options worked fine, but I gave the second one a preference as I find it cleaner. I have now re-implemented both repos using that and everything seems to work just as expected and fine.